docs(quinn): Revise SendStream::write docs

gretchenfrage opened the pull request:
Aligns with 100 columns, attempts to improve clarity, takes inspiration from tokio::io::AsyncWriteExt docs.

Ralith commented:
Initial bikeshedding:
- Why introduce "resolves to" in addition to "yields"?
- Why replace "cancel-safe" with "cancellation safe"?

djc commented on quinn/src/send_stream.rs:
```diff
@@ -43,53 +43,80 @@ impl SendStream {
+    /// the prefix that was written.
     ///
-    /// This operation is cancel-safe.
+    /// This method is cancellation safe: If it does not resolve, no bytes were written.
```
Nit: I don't think we should capitalize after a colon?

flub replied:
nit: tokio uses a `# Cancel safety` heading for these.  Which makes it stand out still a bit more.  I'd be +1 to adopt that convention.

djc commented on quinn/src/send_stream.rs:
```diff
@@ -43,53 +43,80 @@ impl SendStream {
     }
 
-    /// Write chunks to the stream
+    /// Like [`write`](Self::write), except that it writes from a slice of [`Bytes`]
```
I don't like these formulations for the one-line description; I think they're relying too much on indirection. It could be useful to explain the functionality in relation to `Self::write()`, but not in the first line.

djc commented on quinn/src/send_stream.rs:
```diff
@@ -43,53 +43,80 @@ impl SendStream {
+    ///
+    /// This method is *not* cancellation safe: Even if this does not resolve, some prefix of `buf`
+    /// may have been written when previously polled.
+    pub async fn write_all_chunks(&mut self, bufs: &mut [Bytes]) -> Result<(), WriteError> {
```
Nit: don't reorder items in the same commit as you're rewriting the docs.

flub commented:
>     * Why replace "cancel-safe" with "cancellation safe"?

This mirrors how tokio talks about it.  They use a section header named `Cancel safety` and in prose use "This method is cancellation safe".

djc commented:
@gretchenfrage want to polish this up so we can merge it?

gretchenfrage commented:
@djc Yeah, I'd like to prioritize #2168 personally because I don't like how long I've been blocking it. I haven't lost track of anything, I just haven't had capacity for a while.

djc commented:
> @djc Yeah, I'd like to prioritize #2168 personally because I don't like how long I've been blocking it. I haven't lost track of anything, I just haven't had capacity for a while.

Would be nice to get this over the finish line.

gretchenfrage commented:
My revision of this PR is in-progress, I'm hoping I can put it back into review tomorrow. I appreciate your patience with me as my capacity has been reduced a lot recently.

gretchenfrage commented:
- Rebased over main
- Changed the first lines of comments to be imperative, for stylistic consistency with rest of project
- No longer starting first line of docs with "Attempt to" (Tokio sometimes does that but sometimes not. My
  reasoning for removing it is that we don't need to explicitly say that a method which returns a `Result`
  might fail any more than we need to explicitly say that a method which is `async` returns a future, at
  least within the first line of the doc comment.)
- Take flub's suggestion to use Tokio-inspired `# Cancel safety` headers
- Slight tweaks / corrections to my own grammar
- Take Ralith's suggestion to replace phrase "resolves to" with "yields"
- No longer rearranging order of methods
